protocol/cannon-upgrades.md

```diff
@@ -0,0 +1,134 @@
+
+Separating the two upgrades potentially allows cannon MT to ship earlier as it is not blocked by work on 64-bit support.
+However, it means incurring the overhead of governance and coordinating upgrades twice.
+
+If we need to use a new game type, it would invalid user's withdrawals twice.
+
+### Combined Upgrades
+
+Combining both changes into a single upgrade reduces the overhead of governance and upgrade coordination. It ensures
+user's withdrawals are at most invalidated once. If we maintain backwards compatibility in the monorepo, it reduces the
+number of versions being supported from 3 to 2.
+
+## Backwards Compatibility
+
+### New Game Type
+
+The simplest way to maintain backwards compatibility is to deploy the updated cannon VM as a new game type. Changing the
+respected game type in OptimismPortal will invalidate any already proven withdrawals, impacting users.
+
+`op-challenger` will need to be updated to select the version of cannon to use based on the game type. This is
+straight-forward but involves a bunch of mostly boilerplate code.
+
+The legacy version of `cannon` could be made available to `op-challenger` in one of two ways:
+
+1. Retain the ability to build it in the monorepo develop branch.
+
+* `cannon` is automatically included in the `op-challenger` docker image.
+* The `--type` argument to `cannon` can be used to tell it which VM version to run
+
+2. Remove legacy cannon from the monorepo and copy it from an existing docker image when deploying `op-challenger`.
+
+* We will need to make other VMs like `asterisc` available to `op-challenger` in a similar way to this
+* It makes it harder to fix urgent issues in the legacy cannon if required as they would have to be done on a fork
+  from before legacy cannon was removed
+* Has the advantage of removing the legacy code sooner and not needing to make them work in parallel
+
+#### Change Game Type Without Invalidating Withdrawals
+
+The impact to users could be avoided if we upgraded `OptimismPortal` to allow changing the respected game type without
+also invalidating all withdrawals. This is something we would like to do to improve incident response abilities anyway
+but would mean changing the security critical air-gap code. This would need to be audited, delaying deployment.
+
+### Specify Cannon Version in Prestate
+
+Add a flag to the cannon prestate that indicates which version of the VM is required. No new game type is used
+and `op-challenger` will simply provide the appropriate prestate to cannon which then operates in the appropriate mode.
+This makes the upgrade to cannon MT or 64bit work just like any other prestate upgrade.
+
+As the game type isn't changed, users would not need to reprove their withdrawals.
+
+This is really only feasible if we maintain support for legacy cannon in the monorepo until after the upgrade (and any
+existing games are completed). The legacy code could likely shortly after chains have upgraded to the new version. At
```

Comment:
How so? We should be able to do this without keeping support in `develop`, if we cut the final tag of cannon

Reply:
You can't have cannon automatically swap which VM it uses based on the prestate if cannon doesn't have support for the legacy version.  We could in theory do that swap in op-challenger but then it needs even more detailed understanding of cannon's snapshot format and we're trying to go in the other direction and keep the snapshot format isolated in cannon.